OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs

```diff
@@ -75,11 +89,52 @@ public void ResolveOrder(Actor self, Order order)
 				return;
 
 			if (isUpgraded)
-				foreach (var up in info.Upgrades)
-					manager.RevokeUpgrade(self, up, this);
+			{
+				// Play undeploy animation and after that revoke the upgrades
+				self.QueueActivity(false, new CallFunc(() =>
+				{
+					if (!string.IsNullOrEmpty(info.UndeploySound))
+						Sound.Play(info.UndeploySound, self.CenterPosition);
+
+					if (string.IsNullOrEmpty(info.DeployAnimation))
+					{
+						RevokeUpgrades();
+						return;
+					}
+
+					var render = self.TraitOrDefault<ISpriteBody>();
```

Comment:
All of them do Voxel -> Sprite -> Voxel (with Thumpers and GIs being the Sprite -> Sprite exception), so this will actually be good enough for the official mods (custom mods may want Voxel -> Voxel, with a custom MakeAnimation on the deployed voxel, but all of our cases will work.